File: frontend/src/utils/insightsLifestyle.ts
/**
 * Insights lifestyle preferences — defaults, chip rules, request builder.
 * Mapping matches backend/services/wardrobe_gap_context.py.
 */

import {
  INSIGHTS_CLIMATE_OPTIONS,
  INSIGHTS_DRESS_CODE_OPTIONS,
  INSIGHTS_LIFESTYLE_OPTIONS,
  INSIGHTS_STYLE_ACCENT_OPTIONS,
  INSIGHTS_STYLE_PRIMARY_OPTIONS,
} from './constants';

export const MAX_LIFESTYLE_MIX = 3;
export const INSIGHTS_LIFESTYLE_STORAGE_KEY = 'insights_lifestyle_preferences';

export type InsightsLifestyleValue = 'work' | 'everyday' | 'social' | 'formal' | 'sport';
export type InsightsDressCode = 'casual' | 'smart-casual' | 'business-professional' | 'formal';
export type InsightsClimate = 'hot' | 'temperate' | 'cold';
export type InsightsStylePrimary =
  | 'classic'
  | 'smart-casual'
  | 'preppy'
  | 'minimal'
  | 'elegant'
  | 'streetwear'
  | 'sporty';
export type InsightsStyleAccent = 'vintage' | 'edgy' | 'sporty' | 'preppy';

export interface InsightsLifestyleState {
  lifestyleMix: InsightsLifestyleValue[];
  primaryLifestyle: InsightsLifestyleValue;
  dressCodes: InsightsDressCode[];
  climates: InsightsClimate[];
  stylePrimaries: InsightsStylePrimary[];
  styleAccents: InsightsStyleAccent[];
  eventFocus: string | null;
}

/** Old localStorage shape used scalars for dress/climate/style. */
type InsightsLifestyleRaw = Partial<InsightsLifestyleState> & {
  dressCode?: unknown;
  climate?: unknown;
  stylePrimary?: unknown;
  styleAccent?: unknown;
};

const LIFESTYLE_VALUES = INSIGHTS_LIFESTYLE_OPTIONS.map((opt) => opt.value);
const DRESS_CODES = INSIGHTS_DRESS_CODE_OPTIONS.map((opt) => opt.value);
const CLIMATES = INSIGHTS_CLIMATE_OPTIONS.map((opt) => opt.value);
const STYLE_PRIMARIES = INSIGHTS_STYLE_PRIMARY_OPTIONS.map((opt) => opt.value);
const STYLE_ACCENTS = INSIGHTS_STYLE_ACCENT_OPTIONS.map((opt) => opt.value);

const LIFESTYLE_LABELS: Record<string, string> = Object.fromEntries(
  INSIGHTS_LIFESTYLE_OPTIONS.map((opt) => [opt.value, opt.label])
);
const CLIMATE_LABELS: Record<string, string> = Object.fromEntries(
  INSIGHTS_CLIMATE_OPTIONS.map((opt) => [opt.value, opt.label])
);
const STYLE_LABELS: Record<string, string> = {
  ...Object.fromEntries(INSIGHTS_STYLE_PRIMARY_OPTIONS.map((opt) => [opt.value, opt.label])),
  ...Object.fromEntries(INSIGHTS_STYLE_ACCENT_OPTIONS.map((opt) => [opt.value, opt.label])),
};

const OCCASION_FROM_LIFESTYLE: Record<InsightsLifestyleValue, string> = {
  work: 'work',
  everyday: 'everyday',
  social: 'dinner-night-out',
  formal: 'formal-event',
  sport: 'workout',
};

export const DEFAULT_INSIGHTS_LIFESTYLE: InsightsLifestyleState = {
  lifestyleMix: ['work', 'everyday'],
  primaryLifestyle: 'work',
  dressCodes: ['smart-casual'],
  climates: [],
  stylePrimaries: ['classic'],
  styleAccents: [],
  eventFocus: null,
};

function cloneDefaults(): InsightsLifestyleState {
  return {
    ...DEFAULT_INSIGHTS_LIFESTYLE,
    lifestyleMix: [...DEFAULT_INSIGHTS_LIFESTYLE.lifestyleMix],
    dressCodes: [...DEFAULT_INSIGHTS_LIFESTYLE.dressCodes],
    climates: [...DEFAULT_INSIGHTS_LIFESTYLE.climates],
    stylePrimaries: [...DEFAULT_INSIGHTS_LIFESTYLE.stylePrimaries],
    styleAccents: [...DEFAULT_INSIGHTS_LIFESTYLE.styleAccents],
  };
}

function clean(value: string | null | undefined): string | null {
  if (value == null) return null;
  const text = String(value).trim().toLowerCase();
  return text || null;
}

function isLifestyleValue(value: string): value is InsightsLifestyleValue {
  return (LIFESTYLE_VALUES as string[]).includes(value);
}

function isDressCode(value: string): value is InsightsDressCode {
  return (DRESS_CODES as string[]).includes(value);
}

function isClimate(value: string): value is InsightsClimate {
  return (CLIMATES as string[]).includes(value);
}

function isStylePrimary(value: string): value is InsightsStylePrimary {
  return (STYLE_PRIMARIES as string[]).includes(value);
}

function isStyleAccent(value: string): value is InsightsStyleAccent {
  return (STYLE_ACCENTS as string[]).includes(value);
}

function listFromUnknown(value: unknown): string[] {
  if (Array.isArray(value)) {
    const cleaned: string[] = [];
    for (const item of value) {
      const text = clean(item == null ? null : String(item));
      if (text && !cleaned.includes(text)) cleaned.push(text);
    }
    return cleaned;
  }
  const text = clean(value == null ? null : String(value));
  return text ? [text] : [];
}

function pickList(arrayField: unknown, scalarField: unknown): string[] {
  if (Array.isArray(arrayField)) return listFromUnknown(arrayField);
  if (arrayField != null && arrayField !== '') return listFromUnknown(arrayField);
  return listFromUnknown(scalarField);
}

export function lifestyleLabel(value: string): string {
  return LIFESTYLE_LABELS[value] || value;
}

export function dressCodeLabel(value: string): string {
  const match = INSIGHTS_DRESS_CODE_OPTIONS.find((opt) => opt.value === value);
  return match?.label || value;
}

export function normalizeLifestyleMix(
  mix: readonly string[] | null | undefined,
  primary?: string | null
): InsightsLifestyleValue[] {
  const cleaned: InsightsLifestyleValue[] = [];
  for (const item of mix || []) {
    const value = clean(item);
    if (value && isLifestyleValue(value) && !cleaned.includes(value)) {
      cleaned.push(value);
    }
    if (cleaned.length === MAX_LIFESTYLE_MIX) break;
  }

  const primaryValue = clean(primary);
  if (primaryValue && isLifestyleValue(primaryValue)) {
    if (cleaned.includes(primaryValue)) {
      return [primaryValue, ...cleaned.filter((item) => item !== primaryValue)];
    }
    if (cleaned.length < MAX_LIFESTYLE_MIX) {
      return [primaryValue, ...cleaned];
    }
    if (cleaned.length) {
      const next = [...cleaned];
      next[0] = primaryValue;
      return next;
    }
  }

  return cleaned;
}

/** Chip tap rules: max 3, at least 1, one primary. */
export function toggleLifestyleChip(
  mix: readonly InsightsLifestyleValue[],
  primary: InsightsLifestyleValue,
  tapped: InsightsLifestyleValue
): { mix: InsightsLifestyleValue[]; primary: InsightsLifestyleValue } {
  const selected = mix.includes(tapped);

  if (!selected) {
    if (mix.length >= MAX_LIFESTYLE_MIX) {
      return { mix: [...mix], primary };
    }
    return { mix: [...mix, tapped], primary };
  }

  if (tapped !== primary) {
    return { mix: [...mix], primary: tapped };
  }

  if (mix.length <= 1) {
    return { mix: [...mix], primary };
  }

  const nextMix = mix.filter((item) => item !== tapped);
  return { mix: nextMix, primary: nextMix[0] };
}

/** Multi-select; at least one. Tapping the last selected code is a no-op. */
export function toggleDressCodeChip(
  selected: readonly InsightsDressCode[],
  tapped: InsightsDressCode
): InsightsDressCode[] {
  if (!selected.includes(tapped)) {
    return [...selected, tapped];
  }
  if (selected.length <= 1) {
    return [...selected];
  }
  return selected.filter((item) => item !== tapped);
}

/** Climate chips only (Year-round is always on in the UI). Tap selected to deselect. */
export function toggleClimateChip(
  selected: readonly InsightsClimate[],
  tapped: InsightsClimate
): InsightsClimate[] {
  if (!selected.includes(tapped)) {
    return [...selected, tapped];
  }
  return selected.filter((item) => item !== tapped);
}

/**
 * Same tap rules as lifestyle mix, no max: add; tap non-primary to make primary ([0]);
 * tap primary deselects only if another remains.
 */
export function toggleStylePrimaryChip(
  selected: readonly InsightsStylePrimary[],
  tapped: InsightsStylePrimary
): InsightsStylePrimary[] {
  const isSelected = selected.includes(tapped);
  const primary = selected[0];

  if (!isSelected) {
    return [...selected, tapped];
  }

  if (tapped !== primary) {
    return [tapped, ...selected.filter((item) => item !== tapped)];
  }

  if (selected.length <= 1) {
    return [...selected];
  }

  return selected.filter((item) => item !== tapped);
}

/** Multi-select accents. Empty list is None. */
export function toggleStyleAccentChip(
  selected: readonly InsightsStyleAccent[],
  tapped: InsightsStyleAccent
): InsightsStyleAccent[] {
  if (!selected.includes(tapped)) {
    return [...selected, tapped];
  }
  return selected.filter((item) => item !== tapped);
}

export function canonicalOccasion(primary: string, dressCodes: readonly string[]): string {
  const hasBusinessCode = dressCodes.some(
    (code) => code === 'business-professional' || code === 'formal'
  );
  if (primary === 'work' && hasBusinessCode) {
    return 'business';
  }
  return OCCASION_FROM_LIFESTYLE[primary as InsightsLifestyleValue] || 'work';
}

export function canonicalSeason(climates: readonly string[]): string {
  if (climates.length === 1 && climates[0] === 'hot') return 'summer';
  if (climates.length === 1 && climates[0] === 'cold') return 'winter';
  return 'all-season';
}

export function canonicalStyle(stylePrimaries: readonly string[]): string {
  return stylePrimaries[0] || 'classic';
}

export function displayOccasion(mix: readonly string[]): string {
  const labels = mix.map((item) => LIFESTYLE_LABELS[item] || item);
  return labels.length ? labels.join(' + ') : 'Everyday';
}

export function displaySeason(climates: readonly string[]): string {
  const labels = climates.map((item) => CLIMATE_LABELS[item] || item).filter(Boolean);
  if (labels.length) {
    return `Year-round / ${labels.join(' + ')}`;
  }
  return 'Year-round';
}

export function displayStyle(
  stylePrimaries: readonly string[],
  styleAccents: readonly string[] = []
): string {
  const primaryLabels = stylePrimaries.map((item) => STYLE_LABELS[item] || item);
  const primaryText = primaryLabels.length ? primaryLabels.join(' + ') : 'Classic';
  const accentLabels = styleAccents
    .filter((item) => STYLE_ACCENTS.includes(item as InsightsStyleAccent))
    .map((item) => STYLE_LABELS[item] || item);
  if (accentLabels.length) {
    return `${primaryText} with ${accentLabels.join(' + ')} accent`;
  }
  return primaryText;
}

export function sanitizeInsightsLifestyle(
  raw: InsightsLifestyleRaw | null | undefined
): InsightsLifestyleState {
  const mix = normalizeLifestyleMix(raw?.lifestyleMix, raw?.primaryLifestyle);
  const resolvedMix = mix.length ? mix : [...DEFAULT_INSIGHTS_LIFESTYLE.lifestyleMix];

  const dressResolved = pickList(raw?.dressCodes, raw?.dressCode).filter(isDressCode);
  const dressCodes = dressResolved.length
    ? dressResolved
    : [...DEFAULT_INSIGHTS_LIFESTYLE.dressCodes];

  const climates = pickList(raw?.climates, raw?.climate).filter(isClimate);

  const styleResolved = pickList(raw?.stylePrimaries, raw?.stylePrimary).filter(isStylePrimary);
  const stylePrimaries = styleResolved.length
    ? styleResolved
    : [...DEFAULT_INSIGHTS_LIFESTYLE.stylePrimaries];

  const styleAccents = pickList(raw?.styleAccents, raw?.styleAccent).filter(isStyleAccent);
  const event = clean(raw?.eventFocus ?? null);

  return {
    lifestyleMix: resolvedMix,
    primaryLifestyle: resolvedMix[0],
    dressCodes,
    climates,
    stylePrimaries,
    styleAccents,
    eventFocus: event,
  };
}

export function loadInsightsLifestyle(): InsightsLifestyleState {
  try {
    const stored = localStorage.getItem(INSIGHTS_LIFESTYLE_STORAGE_KEY);
    if (!stored) return cloneDefaults();
    return sanitizeInsightsLifestyle(JSON.parse(stored) as InsightsLifestyleRaw);
  } catch {
    return cloneDefaults();
  }
}

export function saveInsightsLifestyle(state: InsightsLifestyleState): void {
  const sanitized = sanitizeInsightsLifestyle(state);
  localStorage.setItem(INSIGHTS_LIFESTYLE_STORAGE_KEY, JSON.stringify(sanitized));
}

export function resetInsightsLifestyle(): InsightsLifestyleState {
  const defaults = cloneDefaults();
  saveInsightsLifestyle(defaults);
  return defaults;
}

export function buildInsightsAnalyzePayload(
  state: InsightsLifestyleState,
  textInput: string
) {
  const sanitized = sanitizeInsightsLifestyle(state);
  const mix = sanitized.lifestyleMix;
  const primary = sanitized.primaryLifestyle;
  const occasion = canonicalOccasion(primary, sanitized.dressCodes);
  const season = canonicalSeason(sanitized.climates);
  const style = canonicalStyle(sanitized.stylePrimaries);

  return {
    occasion,
    season,
    style,
    text_input: textInput,
    lifestyle_mix: mix,
    primary_lifestyle: primary,
    dress_code: sanitized.dressCodes,
    climate: sanitized.climates,
    style_primary: sanitized.stylePrimaries,
    style_accent: sanitized.styleAccents,
    event_focus: sanitized.eventFocus,
  };
}

File: ios-client/OutfitSuggestor/Utils/InsightsLifestyle.swift
//
//  InsightsLifestyle.swift
//  OutfitSuggestor
//
//  Insights-only lifestyle mix, dress code, climate, and style.
//  Mapping matches backend/services/wardrobe_gap_context.py.
//

import Foundation

struct InsightsLifestyleOption: Identifiable, Equatable {
    let value: String
    let label: String
    var id: String { value }
}

struct InsightsLifestyle: Equatable {
    static let maxMix = 3

    static let mixOptions: [InsightsLifestyleOption] = [
        .init(value: "work", label: "Work"),
        .init(value: "everyday", label: "Everyday"),
        .init(value: "social", label: "Social / Dinner"),
        .init(value: "formal", label: "Formal"),
        .init(value: "sport", label: "Sport / Outdoor")
    ]

    static let dressCodeOptions: [InsightsLifestyleOption] = [
        .init(value: "casual", label: "Casual"),
        .init(value: "smart-casual", label: "Smart casual"),
        .init(value: "business-professional", label: "Business professional"),
        .init(value: "formal", label: "Formal")
    ]

    static let climateOptions: [InsightsLifestyleOption] = [
        .init(value: "hot", label: "Hot"),
        .init(value: "temperate", label: "Temperate"),
        .init(value: "cold", label: "Cold")
    ]

    /// Insights primaries — Boho / Romantic / Trendy are not offered.
    static let stylePrimaryOptions: [InsightsLifestyleOption] = [
        .init(value: "classic", label: "Classic"),
        .init(value: "smart-casual", label: "Smart Casual"),
        .init(value: "preppy", label: "Preppy"),
        .init(value: "minimal", label: "Minimal"),
        .init(value: "elegant", label: "Elegant"),
        .init(value: "streetwear", label: "Streetwear"),
        .init(value: "sporty", label: "Sporty")
    ]

    static let styleAccentOptions: [InsightsLifestyleOption] = [
        .init(value: "vintage", label: "Vintage"),
        .init(value: "edgy", label: "Edgy"),
        .init(value: "sporty", label: "Sporty"),
        .init(value: "preppy", label: "Preppy")
    ]

    static let hiddenPrimaryStyles: Set<String> = ["boho", "romantic", "trendy"]

    private static let lifestyleValues = Set(mixOptions.map(\.value))
    private static let dressCodeValues = Set(dressCodeOptions.map(\.value))
    private static let climateValues = Set(climateOptions.map(\.value))
    private static let stylePrimaryValues = Set(stylePrimaryOptions.map(\.value))
    private static let styleAccentValues = Set(styleAccentOptions.map(\.value))

    private static let occasionFromLifestyle: [String: String] = [
        "work": "work",
        "everyday": "everyday",
        "social": "dinner-night-out",
        "formal": "formal-event",
        "sport": "workout"
    ]

    /// Primary lifestyle is always `mix[0]`. Default: Work (primary) + Everyday.
    var mix: [String]
    /// At least one. Default: Smart casual.
    var dressCodes: [String]
    /// Optional climates. Year-round is always on in the UI. Default: none.
    var climates: [String]
    /// Primary style is always `stylePrimaries[0]`. Default: Classic.
    var stylePrimaries: [String]
    /// Empty means None. Default: none.
    var styleAccents: [String]
    var eventFocus: String?

    static let `default` = InsightsLifestyle(
        mix: ["work", "everyday"],
        dressCodes: ["smart-casual"],
        climates: [],
        stylePrimaries: ["classic"],
        styleAccents: [],
        eventFocus: nil
    )

    var primaryLifestyle: String { mix.first ?? "work" }

    var primaryStyle: String { stylePrimaries.first ?? "classic" }

    var canAddMixChip: Bool { mix.count < Self.maxMix }

    var hasNoAccents: Bool { styleAccents.isEmpty }

    func isMixSelected(_ value: String) -> Bool {
        mix.contains(value)
    }

    func isPrimary(_ value: String) -> Bool {
        primaryLifestyle == value
    }

    func isDressCodeSelected(_ value: String) -> Bool {
        dressCodes.contains(value)
    }

    func isClimateSelected(_ value: String) -> Bool {
        climates.contains(value)
    }

    func isStylePrimarySelected(_ value: String) -> Bool {
        stylePrimaries.contains(value)
    }

    func isPrimaryStyle(_ value: String) -> Bool {
        primaryStyle == value
    }

    func isAccentSelected(_ value: String) -> Bool {
        styleAccents.contains(value)
    }

    /// Tap unselected → add (if under max).
    /// Tap selected non-primary → make primary.
    /// Tap primary → deselect only if another chip remains (that other becomes primary).
    mutating func tapMix(_ value: String) {
        guard Self.lifestyleValues.contains(value) else { return }
        var list = mix
        tapPrimaryList(&list, value: value, maxCount: Self.maxMix)
        mix = list
    }

    /// Multi-select. At least one required — last selected code stays.
    mutating func tapDressCode(_ value: String) {
        guard Self.dressCodeValues.contains(value) else { return }
        if let index = dressCodes.firstIndex(of: value) {
            guard dressCodes.count > 1 else { return }
            dressCodes.remove(at: index)
        } else {
            dressCodes.append(value)
        }
    }

    /// Multi-select climates. Tapping a selected climate deselects it. Year-round stays on.
    mutating func tapClimate(_ value: String) {
        guard Self.climateValues.contains(value) else { return }
        if let index = climates.firstIndex(of: value) {
            climates.remove(at: index)
        } else {
            climates.append(value)
        }
    }

    /// Same tap rules as lifestyle mix, with no max.
    mutating func tapStylePrimary(_ value: String) {
        guard Self.stylePrimaryValues.contains(value) else { return }
        var list = stylePrimaries
        tapPrimaryList(&list, value: value, maxCount: nil)
        stylePrimaries = list
    }

    /// Selecting an accent deselects None. Tapping a selected accent deselects it.
    mutating func tapAccent(_ value: String) {
        guard Self.styleAccentValues.contains(value) else { return }
        if let index = styleAccents.firstIndex(of: value) {
            styleAccents.remove(at: index)
        } else {
            styleAccents.append(value)
        }
    }

    /// None clears all accents.
    mutating func clearAccents() {
        styleAccents = []
    }

    func makeRequest(textInput: String, analysisMode: String) -> WardrobeGapAnalysisRequest {
        let mix = normalizedMix()
        let primary = mix.first ?? "work"
        let codes = normalizedDressCodes()
        let climateValues = normalizedClimates()
        let primaries = normalizedStylePrimaries()
        let accents = normalizedStyleAccents()
        let event = eventFocus.flatMap { $0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : $0 }

        return WardrobeGapAnalysisRequest(
            occasion: Self.canonicalOccasion(primary: primary, dressCodes: codes),
            season: Self.canonicalSeason(climates: climateValues),
            style: Self.canonicalStyle(stylePrimary: primaries.first ?? "classic"),
            text_input: textInput,
            analysis_mode: analysisMode,
            lifestyle_mix: mix,
            primary_lifestyle: primary,
            dress_code: codes,
            climate: climateValues.isEmpty ? nil : climateValues,
            style_primary: primaries,
            style_accent: accents.isEmpty ? nil : accents,
            event_focus: event
        )
    }

    func normalizedMix() -> [String] {
        uniqueAllowed(mix, in: Self.lifestyleValues, fallback: ["work", "everyday"], maxCount: Self.maxMix)
    }

    func normalizedDressCodes() -> [String] {
        uniqueAllowed(dressCodes, in: Self.dressCodeValues, fallback: ["smart-casual"], maxCount: nil)
    }

    func normalizedClimates() -> [String] {
        uniqueAllowed(climates, in: Self.climateValues, fallback: [], maxCount: nil)
    }

    func normalizedStylePrimaries() -> [String] {
        uniqueAllowed(stylePrimaries, in: Self.stylePrimaryValues, fallback: ["classic"], maxCount: nil)
    }

    func normalizedStyleAccents() -> [String] {
        uniqueAllowed(styleAccents, in: Self.styleAccentValues, fallback: [], maxCount: nil)
    }

    static func canonicalOccasion(primary: String, dressCodes: [String]) -> String {
        let hasBusinessCode = dressCodes.contains { $0 == "business-professional" || $0 == "formal" }
        if primary == "work" && hasBusinessCode {
            return "business"
        }
        return occasionFromLifestyle[primary] ?? "work"
    }

    static func canonicalSeason(climates: [String]) -> String {
        if climates.count == 1 {
            switch climates[0] {
            case "hot": return "summer"
            case "cold": return "winter"
            default: break
            }
        }
        return "all-season"
    }

    static func canonicalStyle(stylePrimary: String) -> String {
        stylePrimary.isEmpty ? "classic" : stylePrimary
    }

    static func mixLabel(for value: String) -> String {
        mixOptions.first { $0.value == value }?.label ?? value.capitalized
    }

    static func dressCodeLabel(for value: String) -> String {
        dressCodeOptions.first { $0.value == value }?.label ?? value
    }

    static func climateLabel(for value: String) -> String {
        climateOptions.first { $0.value == value }?.label ?? value.capitalized
    }

    static func styleLabel(for value: String) -> String {
        stylePrimaryOptions.first { $0.value == value }?.label
            ?? styleAccentOptions.first { $0.value == value }?.label
            ?? value.capitalized
    }

    /// Shared Primary-badge list: tap unselected to add; tap selected non-primary to make primary;
    /// tap primary deselects only if another remains.
    private mutating func tapPrimaryList(_ list: inout [String], value: String, maxCount: Int?) {
        if let index = list.firstIndex(of: value) {
            if index == 0 {
                guard list.count > 1 else { return }
                list.removeFirst()
            } else {
                list.remove(at: index)
                list.insert(value, at: 0)
            }
        } else {
            if let maxCount, list.count >= maxCount { return }
            list.append(value)
        }
    }

    private func uniqueAllowed(
        _ values: [String],
        in allowed: Set<String>,
        fallback: [String],
        maxCount: Int?
    ) -> [String] {
        var cleaned: [String] = []
        for item in values where allowed.contains(item) && !cleaned.contains(item) {
            cleaned.append(item)
            if let maxCount, cleaned.count == maxCount { break }
        }
        return cleaned.isEmpty ? fallback : cleaned
    }
}
